### PR TITLE
Upgrade precommits

### PR DESCRIPTION
**Pull request by lorenzomammana**
## Summary

This PR aims to upgrade the pre-commit configuration to use the new tools and standards of the Orobix company. It also upgrades mlflow to avoid using an unsafe dependency.

#### Updated

- Remove black from pre-commit hooks
- Use ruff as the main formatter and linting tool
- Upgrade mypy version
- Upgrade mlflow version
- Apply new pre-commits to all files
- Update most of typing to py310 style

## Type of Change

Please select the one relevant option below:

- Refactor

## Checklist

Please confirm that the following tasks have been completed:

- [X] I have tested my changes locally and they work as expected. (Please describe the tests you performed.)
- [X] I have added unit tests for my changes, or updated existing tests if necessary.
- [X] I have updated the documentation, if applicable.
- [X] I have installed pre-commit and run locally for my code changes.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,17 @@
...
 All notable changes to this project will be documented in this file.
 
+### [2.1.4]
+
+#### Updated
+
+- Remove black from pre-commit hooks
+- Use ruff as the main formatter and linting tool
+- Upgrade mypy version
+- Upgrade mlflow version
```
Comment (rcmalli) [inline CHANGELOG.md:12]: 💯

File: quadra/datasets/segmentation.py [hunk 2]
```diff
@@ -82,14 +85,13 @@ def __getitem__(self, index):
...
             mask = cv2.imread(str(mask_path), 0)
-        if self.defect_transform is not None:
-            if label == 1 and np.sum(mask) == 0:
-                if object_mask is not None:
-                    object_mask *= 255
-                aug = self.defect_transform(image=image, mask=mask, object_mask=object_mask, label=label)
-                image = aug["image"]
-                mask = aug["mask"]
-                label = aug["label"]
+        if self.defect_transform is not None and label == 1 and np.sum(mask) == 0:
```
Comment (rcmalli) [inline quadra/datasets/segmentation.py:88]: was it mypy or ruff?

File: quadra/utils/patch/metrics.py [hunk 3]
```diff
@@ -200,8 +203,8 @@ def compute_patch_metrics(
             output_mask = (output_mask > 0).astype(np.uint8)
             gt_img = label(gt_img)
 
-            for i in range(1, gt_img.max() + 1):
-                region = (gt_img == i).astype(bool)
+            for i in range(1, gt_img.max() + 1):  # type: ignore[union-attr]
```
Comment (rcmalli) [inline quadra/utils/patch/metrics.py:206]: before here I think we have need check if 'gt_img' is not None and it is a numpy array. What do you think?

File: quadra/datasets/anomaly.py [hunk 4]
```diff
@@ -213,19 +214,19 @@ def __init__(
...
         self.crop_area = crop_area
-        self.valid_area_mask: Optional[np.ndarray] = None
+        self.valid_area_mask: np.ndarray | None = None
 
         if valid_area_mask is not None:
             if not os.path.exists(valid_area_mask):
                 raise RuntimeError(f"Valid area mask {valid_area_mask} does not exist.")
 
-            self.valid_area_mask = cv2.imread(valid_area_mask, 0) > 0
+            self.valid_area_mask = cv2.imread(valid_area_mask, 0) > 0  # type: ignore[operator]
```
Comment (rcmalli) [inline quadra/datasets/anomaly.py:223]: I think the problem here is that 'cv2.imread()' can still return None if the path is not available. I have seen this multiple times. We can make sure that we found the image path correctly.

**Comment by rcmalli** [comment]
Mostly, It looks fine for me. I have added some comments. One general question: Do we still need the 'pylint' after the 'ruff' update? Did you check any cases where 'ruff' is not enough?

File: quadra/datasets/segmentation.py (hunk 2 above)
Comment (lorenzomammana) [reply quadra/datasets/segmentation.py:88]: This is ruff complaining that you should not write nested if when you can use and

File: quadra/datasets/anomaly.py (hunk 4 above)
Comment (lorenzomammana) [reply quadra/datasets/anomaly.py:223]: Makes sense

**Comment by lorenzomammana** [comment]
> Mostly, It looks fine for me. I have added some comments. One general question: Do we still need the `pylint` after the `ruff` update? Did you check any cases where `ruff` is not enough?

I had a few things that only pylint catched, but I honestly don't remember what they were, for sure ruff is doing most of the work already now

**Pull request merged**